data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/buffer/KafkaBuffer.java

```diff
@@ -0,0 +1,91 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+@DataPrepperPlugin(name = "kafka_buffer", pluginType = Buffer.class, pluginConfigurationType = KafkaBufferConfig.class)
+public class KafkaBuffer<T extends Record<?>> extends AbstractBuffer<T> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(KafkaBuffer.class);
+    private final KafkaCustomProducer producer;
+    private final AbstractBuffer innerBuffer;
+    private final ExecutorService executorService;
+
+    @DataPrepperPluginConstructor
+    public KafkaBuffer(final PluginSetting pluginSetting, final KafkaBufferConfig kafkaBufferConfig, final PluginFactory pluginFactory,
+                       final AcknowledgementSetManager acknowledgementSetManager, final PluginMetrics pluginMetrics){
+        super(pluginSetting);
+        final KafkaCustomProducerFactory kafkaCustomProducerFactory = new KafkaCustomProducerFactory();
+        producer = kafkaCustomProducerFactory.createProducer(kafkaBufferConfig, pluginFactory, pluginSetting,  null, null);
+        final KafkaCustomConsumerFactory kafkaCustomConsumerFactory = new KafkaCustomConsumerFactory();
+        innerBuffer = new BlockingBuffer<>(1000000, 250000, pluginSetting.getPipelineName());
+        final List<KafkaCustomConsumer> consumers = kafkaCustomConsumerFactory.createConsumersForTopic(kafkaBufferConfig, kafkaBufferConfig.getTopic(),
+            innerBuffer, pluginMetrics, acknowledgementSetManager, new AtomicBoolean(false));
+        this.executorService = Executors.newFixedThreadPool(consumers.size());
+        consumers.forEach(this.executorService::submit);
+    }
+
+    @Override
+    public void doWrite(T record, int timeoutInMillis) throws TimeoutException {
+        try {
+            producer.produceRecords(record);
+        } catch (final Exception e) {
+            LOG.error(e.getMessage(), e);
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void doWriteAll(Collection<T> records, int timeoutInMillis) throws Exception {
+        for ( T record: records ) {
+            doWrite(record, timeoutInMillis);
+        }
+    }
+
+    @Override
+    public Map.Entry<Collection<T>, CheckpointState> doRead(int timeoutInMillis) {
+        innerBuffer.doRead(timeoutInMillis);
+
+
+        return innerBuffer.doRead(timeoutInMillis);
+    }
+
+    @Override
+    public void postProcess(final Long recordsInBuffer) {
+        innerBuffer.postProcess(recordsInBuffer);
+    }
+
+    @Override
+    public void doCheckpoint(CheckpointState checkpointState) {
+        innerBuffer.doCheckpoint(checkpointState);
+    }
+
+    @Override
+    public boolean isEmpty() {
```

Comment:
This method should also check that the topic is empty as it is used for ensuring all data is flushed on shutdown

Reply:
This can be addressed in a later PR